rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala

```diff
@@ -0,0 +1,273 @@
+      case _ :: tail => Some(before ++ tail)
+    }
+  }
+
+  @tailrec
+  def possiblyFind[T, R](prop: (T => Option[R]), haystack: List[T]): Option[R] =
+    haystack match {
+      case Nil => None
+      case head :: rest =>
+        prop(head) match {
+          case None    => possiblyFind(prop, rest)
+          case success => success
+        }
+    }
+
+  @tailrec
+  def exists[T](prop: (T => Boolean), haystack: List[T]): Boolean =
+    haystack match {
+      case Nil => false
+      case head :: rest =>
+        if (prop(head))
+          true
+        else
+          exists(prop, rest)
+    }
+
+  def emptyMap: Map[Int, Par] = Map()
+
+  def unLift[F[_], S, A](raw: StateT[F, S, A])(implicit F: Monad[F]): StateT[F, S, F[(S, A)]] =
+    StateT(s => F.pure((s, raw.run(s))))
+
+  def singleOut[A](vals: List[A]): List[(List[A], A, List[A])] =
+    vals.tails
+      .foldLeft((List[A](), List[(List[A], A, List[A])]())) {
+        case ((head, singled), Nil)          => (head, singled)
+        case ((head, singled), elem :: tail) => (elem :: head, (head, elem, tail) :: singled)
+      }
+      ._2
+      .reverse
+
+  // This helper function is useful in several productions
+  def foldMatch(tlist: List[Par], plist: List[Par]): OptionalFreeMap[Unit] =
+    (tlist, plist) match {
+      case (Nil, Nil) => StateT.pure(Unit)
+      case (Nil, _)   => StateT.liftF(None)
+      case (_, Nil)   => StateT.liftF(None)
+      case (t :: trem, p :: prem) => {
+        spatialMatch(t, p).flatMap(_ => foldMatch(trem, prem))
+      }
+    }
+
+  def listMatchSingle[T <: HasLocallyFree](tlist: List[T],
+                                           plist: List[T],
+                                           matcher: (T, T) => OptionalFreeMap[Unit],
+                                           merger: (Par, T) => Par,
+                                           varLevel: Option[Int],
+                                           wildcard: Boolean): OptionalFreeMap[Unit] = {
+    val exactMatch = !wildcard && !varLevel.isEmpty
+    val plen       = plist.length
+    val tlen       = tlist.length
+    if (exactMatch && plen != tlen)
+      StateT.liftF(None)
+    else if (plen > tlen)
+      StateT.liftF(None)
+    // This boundary is very similar to Oleg's once.
+    StateT((s: FreeMap) => {
+      listMatch(tlist, plist, matcher, merger, varLevel, wildcard).run(s) match {
+        case Stream.Empty =>
+          None
+        case head #:: _ =>
+          Some(head)
+      }
+    })
+  }
+
+  def listMatch[T <: HasLocallyFree](tlist: List[T],
+                                     plist: List[T],
+                                     matcher: (T, T) => OptionalFreeMap[Unit],
+                                     merger: (Par, T) => Par,
+                                     varLevel: Option[Int],
+                                     wildcard: Boolean): NonDetFreeMap[Unit] =
+    (tlist, plist) match {
+      // Handle the remainder.
+      case (rem, Nil) =>
```

Comment:
consider a guard here instead of multi-level pattern matching?

Reply:
I think the guard would be worse actually.